### PR TITLE
Fix some edge cases for compile time operations of % operator

### LINKED ISSUE #1562: % operator gives incorrect results compared to Cpython
Got this while testing the % operator alongside `math.mod ` in Lpython - 

```
def f():
    print((8)%3)
    print((8)%(3))
    print(-(8)%3)          #wrong
    print((8)%-3)          #wrong
    print(-(8)%-3)
    print((-8)%3)          #wrong
    print(-8%3)            #wrong
    
f()

(lp) C:\Users\kunni\lpython>python try.py
2
2
1
-1
-2
1
1

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
2
2
-2
2
-2
-2
-2
```

### PR DESCRIPTION
**Pull request by faze-geek**
Fixes #1562 
On master - 

```
def f():
    print((8)%3)
    print((8)%-3)         
    print((-8)%3)
    print((8.0)%3.0)
    print((8.0)%-3.0)         
    print((-8.0)%3.0)            
    
f()

(lp) C:\Users\kunni\lpython>python try.py
2
-1
1
2.0
-1.0
1.0

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
2
2
-2
2.00000000000000000e+00
2.00000000000000000e+00
-2.00000000000000000e+00
```

On branch - 
```
def f():
    print((8)%3)
    print((8)%-3)         
    print((-8)%3)
    print((8.0)%3.0)
    print((8.0)%-3.0)         
    print((-8.0)%3.0)            
    
f()

(lp) C:\Users\kunni\lpython>python try.py
2
-1
1
2.0
-1.0
1.0

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
2
-1
1
2.00000000000000000e+00
-1.00000000000000000e+00
1.00000000000000000e+00
```

### CODE REVIEW

**Comment by faze-geek** [comment]
On master -
```
def f():
    print(8%0)
    
f()

(lp) C:\Users\kunni\lpython>python try.py
Traceback (most recent call last):
ZeroDivisionError: integer division or modulo by zero

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
Floating point exception
```
On branch -
```
def f():
    print((8)%0)
    
f()
(lp) C:\Users\kunni\lpython>python try.py
Traceback (most recent call last):
ZeroDivisionError: integer division or modulo by zero

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
semantic error: Integer division or modulo by zero
 --> try.py:2:11
  |
2 |     print((8)%0)
  |           ^^^^^
```

File: src/lpython/semantics/python_comptime_eval.h [hunk 1]
```diff
@@ -211,13 +211,19 @@ struct PythonIntrinsicProcedures {
         if (ASRUtils::is_integer(*type)) {
             int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(arg1)->m_n;
             int64_t b = ASR::down_cast<ASR::IntegerConstant_t>(arg2)->m_n;
+            if(b == 0){
```
Comment (czgdp1807) [inline src/lpython/semantics/python_comptime_eval.h:214]:
```suggestion
            if(b == 0) {
```

File: src/lpython/semantics/python_comptime_eval.h [hunk 2]
```diff
@@ -211,13 +211,19 @@ struct PythonIntrinsicProcedures {
...
+                throw SemanticError("Integer division or modulo by zero",loc);
+            }
             return ASR::down_cast<ASR::expr_t>(
-                ASR::make_IntegerConstant_t(al, loc, a%b, type));
+                ASR::make_IntegerConstant_t(al, loc, ((a%b)+b)%b, type));
```
Comment (czgdp1807) [inline src/lpython/semantics/python_comptime_eval.h:218]: So you are trying to handle negative and positive combinations through this, correct? What about doing, `a > 0 && b > 0`, `a < 0 && b < 0`, `a > 0 && b < 0`, `a < 0 && b > 0` and `a == 0` in an if-else-if ladder. Right now the formula seems very cryptic. Or if you have a reference from where you have taken it to the formula then add it.

File: src/lpython/semantics/python_comptime_eval.h [hunk 3]
```diff
@@ -211,13 +211,19 @@ struct PythonIntrinsicProcedures {
...
         } else if (ASRUtils::is_real(*type)) {
             double a = ASR::down_cast<ASR::RealConstant_t>(arg1)->m_r;
             double b = ASR::down_cast<ASR::RealConstant_t>(arg2)->m_r;
+            if (b == double(0)) {
```
Comment (czgdp1807) [inline src/lpython/semantics/python_comptime_eval.h:222]: Doing `double(0)` doesn't matter here I guess.

File: src/lpython/semantics/python_comptime_eval.h [hunk 4]
```diff
@@ -211,13 +211,19 @@ struct PythonIntrinsicProcedures {
...
+                throw SemanticError("Float division or modulo by zero", loc);
+            }
             return ASR::down_cast<ASR::expr_t>(
-                ASR::make_RealConstant_t(al, loc, std::fmod(a, b), type));
+                ASR::make_RealConstant_t(al, loc, std::fmod(std::fmod(a, b) + b, b), type));
```
Comment (czgdp1807) [inline src/lpython/semantics/python_comptime_eval.h:226]: I guess same formula but with `std::fmod` calls. Will recommend the same here as my previous review for integers.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: src/lpython/semantics/python_comptime_eval.h (hunk 2 above)
Comment (gptsarthak) [reply src/lpython/semantics/python_comptime_eval.h:218]: I believe he used the formula from my comment here: https://github.com/lcompilers/lpython/issues/1562#issuecomment-1456821096
I dry ran different combinations of positive and negative values before writing my comment, and it seems to work fine.

Comment (faze-geek) [reply src/lpython/semantics/python_comptime_eval.h:218]: >.Or if you have a reference from where you have taken it to the formula then add it.

You can refer to this link, It correctly handles all the cases.
https://stackoverflow.com/questions/1907565/c-and-python-different-behaviour-of-the-modulo-operation

Comment (czgdp1807) [reply src/lpython/semantics/python_comptime_eval.h:218]: Please add the above link in your code comments as well.

File: src/lpython/semantics/python_comptime_eval.h (hunk 4 above)
Comment (czgdp1807) [reply src/lpython/semantics/python_comptime_eval.h:226]: N.B. - https://github.com/lcompilers/lpython/pull/1566/files#r1130429810

File: src/lpython/semantics/python_comptime_eval.h (hunk 2 above)
Comment (faze-geek) [reply src/lpython/semantics/python_comptime_eval.h:221]: I have added comments for my changes. You can have a look now .

File: src/lpython/semantics/python_comptime_eval.h (hunk 4 above)
Comment (faze-geek) [reply src/lpython/semantics/python_comptime_eval.h:229]: I guess adding the link as comments once suffices. Do I need to add it above this snippet as well ?

**Comment by faze-geek** [comment]
@czgdp1807 is the latest commit regarding comments fine ?

**Comment by czgdp1807** [comment]
Please mark your PR ready for review whenever you feel its ready.

File: src/lpython/semantics/python_comptime_eval.h (hunk 4 above)
Comment (czgdp1807) [reply src/lpython/semantics/python_comptime_eval.h:229]: Add the link here as well.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Review by czgdp1807 — CHANGES_REQUESTED** [review]
Please add the tests for it as well.

**Comment by faze-geek** [comment]
@Thirumalai-Shaktivel . Could you help me to resolve the conflicts here. Locally I didn't get any conflicts, neither did `python run_tests.py` fail after updating them, so I'm not sure how to handle it.

File: integration_tests/expr_05.py [hunk 5]
```diff
@@ -69,6 +71,11 @@ def main0():
     assert a == 0
     b **= 4
     assert b == 256
+    # Test Issue 1562
+    assert ((-8)%3) == 1
+    assert ((8)%-3) == -1
+    assert abs((11.0%-3.0) - (-1.0)) < eps
+    assert abs((-11.0%3.0) - (1.0)) < eps
```
Comment (czgdp1807) [inline integration_tests/expr_05.py:78]: A case for both less than 0 makes sense?

**Comment by czgdp1807** [comment]
You simply had to,

1. Update your `main` branch locally from https://github.com/lcompilers/lpython
2. Checkout into `mod`.
3. Execute `git merge main`. It will show CONFLICTs in the output on your terminal/command line.
4. Do `git add .` for safety.
5. Do a clean build.
6. Update reference tests via, `python run_tests.py -u`.
7. Do `git add .` and `git commit -m "your_message"`.
8. Push changes here.

File: integration_tests/expr_05.py (hunk 5 above)
Comment (faze-geek) [reply integration_tests/expr_05.py:79]: Yes, I added 1 such test.

**Comment by faze-geek** [comment]
> You simply had to,
> 
> 1. Update your `main` branch locally from https://github.com/lcompilers/lpython
> 2. Checkout into `mod`.
> 3. Execute `git merge main`. It will show CONFLICTs in the output on your terminal/command line.
> 4. Do `git add .` for safety.
> 5. Do a clean build.
> 6. Update reference tests via, `python run_tests.py -u`.
> 7. Do `git add .` and `git commit -m "your_message"`.
> 8. Push changes here.

Thanks, I'll keep this in mind. My apologies on using `reset --hard` on this instance as it manipulates the git history.

**Comment by faze-geek** [comment]
This pull request is now ready for review.

**Review by czgdp1807 — APPROVED** [review]

**Pull request merged**